orion/primitives/tadgan.py

```diff
@@ -82,244 +142,317 @@ def __getstate__(self):
-        x = Input(shape=input_shape)
-        model = keras.models.Sequential()
+    @classmethod
+    def _build_model(cls, hyperparameters: dict, layers: list, input_shape: tuple,
+                     name: str) -> Model:
+        x = tf.keras.layers.Input(shape=input_shape)
+        model = tf.keras.Sequential(name=name)
 
         for layer in layers:
             built_layer = build_layer(layer, hyperparameters)
             model.add(built_layer)
 
         return Model(x, model(x))
 
-    def _setdefault(self, kwargs, key, value):
-        if key in kwargs:
-            return
-
-        if key in self.hyperparameters and self.hyperparameters[key] is None:
-            kwargs[key] = value
-
-    def _wasserstein_loss(self, y_true, y_pred):
-        return K.mean(y_true * y_pred)
-
-    def _gradient_penalty_loss(self, y_true, y_pred, averaged_samples):
-        gradients = K.gradients(y_pred, averaged_samples)[0]
-        gradients_sqr = K.square(gradients)
-        gradients_sqr_sum = K.sum(gradients_sqr, axis=np.arange(1, len(gradients_sqr.shape)))
-        gradient_l2_norm = K.sqrt(gradients_sqr_sum)
-        gradient_penalty = K.square(1 - gradient_l2_norm)
-        return K.mean(gradient_penalty)
-
-    def __init__(self, layers_encoder, layers_generator, layers_critic_x, layers_critic_z,
-                 optimizer, input_shape=(100, 1), target_shape=(100, 1), latent_dim=20,
-                 learning_rate=0.0005, epochs=2000, batch_size=64, iterations_critic=5,
-                 **hyperparameters):
-
-        self.shape = input_shape
-        self.epochs = epochs
-        self.batch_size = batch_size
-        self.latent_dim = latent_dim
-        self.latent_shape = (latent_dim, 1)
-        self.target_shape = target_shape
-        self.iterations_critic = iterations_critic
-
-        self.layers_encoder, self.layers_generator = layers_encoder, layers_generator
-        self.layers_critic_x, self.layers_critic_z = layers_critic_x, layers_critic_z
-
-        self.optimizer = import_object(optimizer)(learning_rate)
-
-        self.hyperparameters = hyperparameters
-        self._fitted = False
-
-    def _augment_hyperparameters(self, X, y, kwargs):
+    def _augment_hyperparameters(self, X: ndarray, y: ndarray, **kwargs) -> dict:
         shape = np.asarray(X)[0].shape
-        length = shape[0]
+        length, input_dim = shape
         target_shape = np.asarray(y)[0].shape
+        output_dim = target_shape[1]
 
-        # to infer the shape
+        # Infers the shape.
         self.shape = self.shape or shape
         self.target_shape = self.target_shape or target_shape
-
-        self._setdefault(kwargs, 'generator_reshape_dim', length // 2)
-        self._setdefault(kwargs, 'generator_reshape_shape', (length // 2, 1))
-        self._setdefault(kwargs, 'encoder_reshape_shape', self.latent_shape)
-
+        self.latent_shape = self.latent_shape or (self.latent_dim, output_dim)
```

Comment:
if you apply my suggestion of keeping `latent_dim` and removing `latent_shape` from the model parameters, this line becomes `self.latent_shape = (self.latent_dim, output_dim)`